docs/proposals/apiserver_watch.md

```diff
@@ -0,0 +1,132 @@
+responsible for watching all the changes of objects of a given resource type.
+However, this watch can potentially expire at any time and reconnecting can
+return "too old resource version". In that case relisting is necessary.
+To avoid LIST requests coming from all the watcher at that time, we will
+introduce an additional etcd event type:
+[EtcdResync](../../pkg/tools/etcd_helper_watch.go#L36)
+
+  Whenever reslisting will be done to refresh the internal watch to etcd,
+  EtcdResync event will be send to all the watchers. It will contain the
+  full list of all the objects the watcher is interested in (appropriately
+  filtered) as the parameter of this watch event.
+  Thus, we need to create the EtcdResync event, extend watch.Interface and
+  its implementations to support it and handle those events appropriately
+  in places like
+  [Reflector](../../pkg/client/cache/reflector.go)
+
+4. Create a FIFO cache with a given capacity. This should be as simple as
+having an array an treating it as a cyclic one. Note that in the cache
+we should store two things: the object watched from etcd and its resource
+version (etcdIndex). Obviously resource versions will be increasing (if
+we treat the array as a cyclic one), but they are necessary for registering
+a new watcher that is interested in all the changes that happened after
+a given etcdIndex (resource version).
+
+  We should also support LIST operation from the beginning and the end of
+  he cache (i.e. from the moment just before the oldest cached watch event
+  and from the moment just after the newest cached watch event). Both of
+  them should be very simple to do - we can incrementally update those list
+  whenever we add/delete a new element to/from cache.
+  This will be necessary to appropriately process EtcdResync watch events
+  (I will refer to it later in this proposal).
+
+5. Create the new implementation of the API, that will internally have a
+single watch open to etcd and will store the data received from etcd in
+the FIFO cache - this includes implementing registration of a new watcher
+which will start a new go-routine responsible for iterating over the cache
+and sending all the objects watcher is interested in (by applying filtering
+function) to the watcher.
+
+  Note that if a per-watcher go-routine will fall behind the oldest element
+  in the cache, this would require sending EtcdResync event to the client
+  and starting its go-routine from that point. Similarly, we can send an
+  EtcdResync event if the watch request comes with resourceVersion older
+  than the oldest element in the cache.
+
+6. Add a support for processing "error too old" from etcd, which will require:
+  - stopping to send any watch events to registered watchers
+  - clearing the cache and relisting all objects from etcd (and storing them
+    in a cleared cache)
+  - sending EtcdResync event to all watchers (appropriately filtering objects
```

Comment:
hmm - Resync is in a way List, but it's send to the client as part of watch stream without forcing him to explicitly ask for it. The reason for that is to allow for "watch window expiration" in the internal watch etcd, but don't force all the clients to relist from etcd at that point.
Does this and the comment above explains what I'm proposing?

[BTW it seems to me that at least partially there is some confusion about the terminology I'm using - in that cases just let my know - I'm happy to change it so that it's more understandable]